### DISCUSSION TITLE
Preview Option for "Ask" and "Explain" Responses

### DISCUSSION THREAD

**Discussion by frettig-radiox** [open-webui/open-webui#6061] [Ideas]
The new **"Ask"** and **"Explain"** features are awsome! However, I would like to suggest a small improvement to further streamline their functionality and keep the chat interface cleaner and more organized.

**Feature Description**:  
Currently, when users select the "Ask" or "Explain" buttons on highlighted text, the resulting responses are immediately added to the conversation history. I propose a modification where, instead of instantly adding the result to the chat, the response is first displayed in a temporary text preview window. This window would allow users to review the response, and they can then choose whether to add it to the conversation by clicking an **"Add to Conversation"** button. If the user doesn’t need to add the response, they can simply close the preview window without affecting the conversation thread.

**Expected Behavior**:  
When a user clicks "Ask" or "Explain," instead of directly appending the result to the chat log, a preview window appears, showing the generated response. The user can then decide if the answer should be included in the conversation by clicking **"Add to Conversation"**, or dismiss it if they find the response unnecessary for the chat. This prevents the conversation from becoming cluttered with every question or explanation request, especially when many inquiries are made in quick succession.

![ask_dialog](https://github.com/user-attachments/assets/bda57901-bfc4-481e-af7d-76c307ad6d14)

Example:
- A user highlights a sentence, clicks "Explain," and the explanation is displayed in a pop-up window. After reviewing the response, the user decides whether to add it to the chat conversation or close the window if it was just for personal reference.

**Technical Requirements**:

- **Temporary Preview Window**:  
  A modal or pop-up text window that displays the response from the LLM (whether it is an "Ask" or "Explain" query). This window should include a button like "Add to Conversation" to append the text to the chat.

- **Dismissal Option**:  
  Users should be able to dismiss the preview window without adding the text to the chat if they don’t need it, preserving the cleanliness of the conversation history.

**UX/UI Changes**:

- Add a text preview window with buttons for "Add to Conversation" and "Close" after an "Ask" or "Explain" request is processed.
- Modify the existing flow so that the response isn’t automatically added to the chat but is shown in the preview first.

### Potential Benefits:

- **Cleaner Chat History**: This improvement would prevent the chat conversation from becoming overcrowded with responses from numerous "Ask" or "Explain" queries, especially in situations where users are exploring or testing multiple pieces of content.
  
- **User Control**: By allowing users to review the LLM’s response before it enters the conversation, they maintain more control over what content is actually added, reducing irrelevant or redundant information in the chat.

- **Efficient Workflow**: Users can still leverage the powerful "Ask" and "Explain" features without worrying about cluttering the chat, leading to a more efficient and organized interaction with the system.

**Comment by davizucon** [comment]
Awesome feature!

If I may suggest a change, there is currently a fixed text for the prompt ([here](https://github.com/open-webui/open-webui/blob/caeca9d300b55d3ce5794e498e5424e72ffae177/src/lib/components/chat/Messages/ResponseMessage.svelte#L499)), it would be nice if we could customize this text or translate it.

Thanks!

**Comment by frettig-radiox** [comment]
do you suggest something like this?
We could make it generic and provide the json as a config option (later) 

[this is also related to this discussion](https://github.com/open-webui/open-webui/discussions/6059)

```json
{
  "context_actions": [
    {
      "name": "ask",
      "input": {
        "type": "text",
        "placeholder": "Enter your question"
      },
      "output": {
        "type": "append", 
        "preview": true,
        "buttons": [true, false]  // "append to conversation": enabled - "replace selection" button disabled
      },
      "llm_prompt": {
        "context_inclusion": true,
        "prompt": "\"\"\"\n${content}\n\"\"\"\n${input}`"
      }
    },
    {
      "name": "explain",
      "input": {
        "type": "none"
      },
      "output": {
        "type": "append", 
        "preview": true,
        "buttons": [true, false]  // "append to conversation": enabled - "replace selection" button disabled
      },
      "llm_prompt": {
        "context_inclusion": true,
        "prompt": "Explain this section to me in more detail\n\n\"\"\"\n${content}\n\"\"\""
      }
    },
    {
      "name": "modify",
      "input": {
        "type": "text",
        "placeholder": "Enter your instructions (e.g. summarize, ...)"
      },
      "output": {
        "type": "replace", 
        "preview": true,
        "buttons": [true, true]  // "append to conversation": enabled - "replace selection" button enabled
      },
      "llm_prompt": {
        "context_inclusion": true,
        "prompt": "Modify the section \n\n\"\"\"\n${content}\n\"\"\" according to these instructions: ${input}"
      }
    }
  ]
}
```

**Comment by tjbck** [comment]
All great ideas, will start investigating shortly.

**Comment by frettig-radiox** [comment]
Applying "actions" to the selection is a cool feature. It would be great to have the selection available in actions and filters api as well!

**Comment by frettig-radiox** [comment]
... and if the existing Actions API is extended with a flag, it could control whether the action button appears at the bottom of the chat (as it is now) or contextually (like the "ask" and "explain" buttons). We could also bundle it with a model or sharing it online like other actions and filters.

**Comment by tjbck** [comment]
![image](https://github.com/user-attachments/assets/b4f3e323-b8d8-48c9-9ad5-5f490c45271a)

Preview option added to dev!

**Comment by frettig-radiox** [comment]
It works perfectly well! However, I have two remarks:

1. **Prompt Template Adjustment**  
   I assume your prompt template looks something like this: 

Explain this section to me in more detail  

`` `
${selection}  
tick_tick_tick 

   If the `selection` contains a code block (which likely starts with ```), the output may not render correctly due to nested code blocks using the same marker string.  

   To avoid this, it would be better to use `~~~` for the outer code block since LLMs tend to generate code blocks with ```.  

   **Solution:**  
   Update the prompt template to:  

   ```  
   Explain this section to me in more detail  

   ~~~  
   ${selection}  
   ~~~  
   ```  
   This ensures that code blocks render correctly, as shown below.


![code_bock](https://github.com/user-attachments/assets/a066617c-fa3c-4ab7-bb34-a8b7dccfc04f)

2. **Button Label Precision**  
   The button labeled **"Add"** could be more descriptive. Suggestions include **"Add to Chat"** or **"Append to Chat"**, which better convey the intended action.

**Comment by astroboylrx** [comment]
These are great features! I wonder does "Explain" respect the default system prompt? My system prompt ask for "concise" answer but "Explain" seems to be quite lengthly.
